CentralUnit/Communication/XbeeNetwork.cs

```diff
@@ -0,0 +1,416 @@
+                    {
+                        Log.Error("Error while parsing packet: ", ex);
+                    }
+                }
+            }
+
+            if (packet == null)
+            {
+                Log.ErrorFormat("Packet in queue was invalid, data: {0}", BitConverter.ToString(buffer));
+            }
+
+            return packet;
+        }
+
+        /// <summary>
+        /// Starts a new packet by allocating a new buffer and resetting the index to 0.
+        /// If index is greater than 0, enqueues the current packet to be processed and distributed among devices.
+        /// </summary>
+        private void StartNewPacket()
+        {
+            if (this.currentPacketPosition > 0)
+            {
+                this.receiveQueue.Enqueue(this.currentPacket);
+            }
+
+            this.currentPacket = new byte[1024];
+            this.currentPacketPosition = 0;
+        }
+
+        /// <summary>
+        /// Handles change in communication channel connection state.
+        /// Notifies all devices in this network of the new connection state.
+        /// </summary>
+        /// <param name="sender">The sender of the event</param>
+        /// <param name="e">EventArgs supplied with the event</param>
+        private void CommunicationChannel_ConnectionStateChanged(object sender, EventArgs e)
+        {
+            foreach (XbeeSerialCommunication xbee in this.devices)
+            {
+                xbee.OnConnectionStateChanged(this.communicationChannel.Connected);
+            }
+        }
+
+        /// <summary>
+        /// Thread main method.
+        /// </summary>
+        private void RunXbeeNetwork()
+        {
+            try
+            {
+                DateTime threadStartTime = DateTime.Now;
+
+                // Wait for up to 10 seconds for a connection. This happens on another thread.
+                while ((!this.communicationChannel.Connected) &&
+                        (DateTime.Now.Subtract(threadStartTime).TotalMilliseconds < 10000))
+                {
+                    Thread.Sleep(100);
+                }
+
+                while (this.communicationChannel.Connected)
+                {
+                    this.ManageTxQueue();
+
+                    lock (this.packetLock)
+                    {
+                        this.ManageRxQueue();
+                    }
+
+                    Thread.Sleep(10);
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Error("Exception on XbeeNetwork thread", ex);
+            }
+        }
+
+        /// <summary>
+        /// Process a received <see cref="XBeePacket"/>
+        /// </summary>
+        /// <param name="packet">The <see cref="XBeePacket"/> to process.</param>
+        private void ProcessXbeePacket(XBeePacket packet)
+        {
+            if (packet != null)
+            {
+                if (packet.GetType() == typeof(ExplicitRxIndicatorPacket))
```

Comment:
there is more convient syntax: `if(packet is ExplicitIndicatorPacket eriPacket) {//do thinhs with eriPacket}`

Reply:
see also line 324